flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobSubmitHandler.java

```diff
@@ -34,38 +40,137 @@
 		super(localRestAddress, leaderRetriever, timeout, headers, JobSubmitHeaders.getInstance());
+		this.executor = executor;
 	}
 
 	@Override
 	protected CompletableFuture<JobSubmitResponseBody> handleRequest(@Nonnull HandlerRequest<JobSubmitRequestBody, EmptyMessageParameters> request, @Nonnull DispatcherGateway gateway) throws RestHandlerException {
-		JobGraph jobGraph;
-		try {
-			ObjectInputStream objectIn = new ObjectInputStream(new ByteArrayInputStream(request.getRequestBody().serializedJobGraph));
-			jobGraph = (JobGraph) objectIn.readObject();
-		} catch (Exception e) {
+		Collection<File> uploadedFiles = request.getUploadedFiles();
+		Map<String, Path> nameToFile = uploadedFiles.stream().collect(Collectors.toMap(
+			path -> path.toPath().getFileName().toString(),
+			File::toPath
+		));
+
+		if (uploadedFiles.size() != nameToFile.size()) {
 			throw new RestHandlerException(
-				"Failed to deserialize JobGraph.",
-				HttpResponseStatus.BAD_REQUEST,
-				e);
+				String.format("The number of uploaded files was %s than the expected count. Expected: %s Actual %s",
+					uploadedFiles.size() < nameToFile.size() ? "lower" : "higher",
+					nameToFile.size(),
+					uploadedFiles.size()),
+				HttpResponseStatus.BAD_REQUEST
+			);
+		}
+
+		JobSubmitRequestBody requestBody = request.getRequestBody();
+
+		Path jobGraphFile = getPathAndAssertUpload(requestBody.jobGraphFileName, FILE_TYPE_GRAPH, nameToFile);
+
+		CompletableFuture<JobGraph> jobGraphFuture = CompletableFuture.supplyAsync(() -> {
+			JobGraph jobGraph;
+			try (ObjectInputStream objectIn = new ObjectInputStream(Files.newInputStream(jobGraphFile))) {
+				jobGraph = (JobGraph) objectIn.readObject();
+			} catch (Exception e) {
+				throw new CompletionException(new RestHandlerException(
+					"Failed to deserialize JobGraph.",
+					HttpResponseStatus.BAD_REQUEST,
+					e));
+			}
+			return jobGraph;
+		}, executor);
+
+		Collection<org.apache.flink.core.fs.Path> jarFiles = getJarFilesToUpload(nameToFile, requestBody.jarFileNames);
+
+		Collection<Tuple2<String, org.apache.flink.core.fs.Path>> artifacts = getArtifactFilesToUpload(nameToFile, requestBody.artifactFileNames);
+
+		CompletableFuture<Integer> blobServerPortFuture = gateway.getBlobServerPort(timeout);
+
+		CompletableFuture<JobGraph> finalizedJobGraphFuture = jobGraphFuture.thenCombine(blobServerPortFuture, (jobGraph, blobServerPort) -> {
+			final InetSocketAddress address = new InetSocketAddress(gateway.getHostname(), blobServerPort);
+			try (BlobClient blobClient = new BlobClient(address, new Configuration())){
+				Collection<PermanentBlobKey> jarBlobKeys = ClientUtils.uploadUserJars(jobGraph.getJobID(), jarFiles, blobClient);
+				ClientUtils.setUserJarBlobKeys(jarBlobKeys, jobGraph);
+				Collection<Tuple2<String, PermanentBlobKey>> artifactBlobKeys = ClientUtils.uploadUserArtifacts(jobGraph.getJobID(), artifacts, blobClient);
+				ClientUtils.setUserArtifactBlobKeys(jobGraph, artifactBlobKeys);
```

Comment:
This code snippet looks very much like `ClientUtils#uploadJobGraphFiles`. Maybe we could refactor it to be reused here.

Reply:
The key difference is that `uploadJobGraphFiles` retrieves the jar/artifact collections to upload from the jobgraph, where as we supply them here directly.

Otherwise we have to modify the JobGraph, or modify `uploadJobGraphFiles` to also accept the collections, but then the method is a bit ambiguous (does it use the JobGraph, the collection, or both?).

Reply:
This makes sense. What about renaming `uploadJobGraphFiles` into `extractAndUploadJobGraphFiles(JobGraph)` which extracts the files to be uploaded from the `JobGraph` and calls `uploadJobGraphFiles(JobGraph, Collection<File> jarFiles)` which takes the `JobGraph` and the file collections which it uploads. Then we could simply call `uploadJobGraphFiles` from the `JobSubmitHandler`.